pkg/session/syssession/session.go

```diff
@@ -0,0 +1,477 @@
+	if err != nil {
+		return err
+	}
+
+	if txn.Valid() {
+		return errors.New("txn is still valid")
+	}
+
+	return nil
+}
+
+// OwnerResetState resets the state of the session by owner
+func (s *session) OwnerResetState(ctx context.Context, caller sessionOwner) error {
+	sctx, exit, err := s.EnterOperation(caller)
+	if err != nil {
+		return err
+	}
+	defer exit()
+	sctx.RollbackTxn(ctx)
+	return nil
+}
+
+// EnterOperation enters an operation on the session.
+// It receives a caller with the type sessionOwner as the argument,
+// and if the session is closed or the caller is not the owner, it will return an error.
+// If the operation is entered successfully,
+// it will return the internal SessionContext for further usage and a function to exit the operation.
+func (s *session) EnterOperation(caller sessionOwner) (SessionContext, func(), error) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	seqStart := s.seqInc()
+	if err := s.checkOwner(caller, "EnterOperation error: "); err != nil {
+		s.reportError(
+			err.Error(),
+			zap.Uint64("seqStart", seqStart),
+			zap.String("caller", objectStr(caller)),
+		)
+		return nil, nil, err
+	}
+
+	s.inuse++
+	exit := false
+	return s.sctx, func() {
+		r := recover()
+		s.mu.Lock()
+		defer s.mu.Unlock()
+		seqEnd := s.seqInc()
+		if exit {
+			s.reportError(
+				"EnterOperation error: multiple exits for the same operation",
+				zap.Uint64("seqStart", seqStart),
+				zap.Uint64("seqEnd", seqEnd),
+				zap.String("caller", objectStr(caller)),
+			)
+			return
+		}
+
+		exit = true
+		s.inuse--
+		if s.owner != caller {
+			s.reportError(
+				"ExitOperation error: session owner transferred when executing operation",
+				zap.Uint64("seqStart", seqStart),
+				zap.Uint64("seqEnd", seqEnd),
+				zap.String("caller", objectStr(caller)),
+			)
+		}
+		if r != nil {
+			logutil.BgLogger().Error(
+				"EnterOperation error: panic occurs, avoid to use session again",
+				zap.Any("panic", r),
+				zap.Uint64("seqStart", seqStart),
+				zap.Uint64("seqEnd", seqEnd),
+				zap.String("sctx", objectStr(s.sctx)),
+				zap.String("caller", objectStr(caller)),
+				zap.String("owner", objectStr(s.owner)),
+				zap.Stack("stack"),
+			)
+			s.avoidReuse = true
+			panic(r)
+		}
+	}, nil
+}
+
+// OwnerClose indicates to close the session.
+// It receives a caller with the type sessionOwner as the argument,
+// and if the caller is not the owner, or it is already closed, it will return without doing anything.
+func (s *session) OwnerClose(caller sessionOwner) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if caller != nil && caller == s.owner {
+		// only close the session when the caller is the owner
+		s.doClose(s.seqInc())
+	}
+}
+
+// Close closes the session without checking the owner.
+func (s *session) Close() {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	s.doClose(s.seqInc())
+}
+
+// doClose closes the session.
+// It should be called with the protection of the mutex.
+func (s *session) doClose(seq uint64) {
+	if s.owner == nil {
```

Comment:
When will the owner be nil, but we still try to close it?

Reply:
It may happen when we call `session.Close()` twice:

```
var s *session
s = ...
// first Close will set `s.owner = nil`
s.Close()
// The second Close will do nothing
s.Close()
```

Reply:
Could you please add some comments to explain it? Thanks!

Reply:
Done